File: app/src/main/kotlin/net/tigr/soulcalc/domain/parser/Token.kt
/*
 * SPDX-License-Identifier: MIT
 * Copyright (c) 2026 Albert Zenkoff
 */

package net.tigr.soulcalc.domain.parser

/**
 * Represents a token produced by the lexer.
 *
 * @property type The type of token
 * @property value The string value of the token
 * @property position The starting position in the input string
 */
data class Token(
    val type: TokenType,
    val value: String,
    val position: Int
)

/**
 * Enumeration of all token types recognized by the lexer.
 */
enum class TokenType {
    // Literals
    NUMBER,         // 123, 45.67

    // Operators
    PLUS,           // +
    MINUS,          // - or −
    MULTIPLY,       // * or ×
    DIVIDE,         // / or ÷
    PERCENT,        // %
    POWER,          // ^
    EQUALS,         // =

    // Delimiters
    LPAREN,         // (
    RPAREN,         // )

    // Identifiers
    VARIABLE,       // $name (named variable)
    LINE_REF,       // $1, $2 (line reference)

    // Functions
    SQRT,           // sqrt

    // Special
    EOF,            // End of input
    ERROR           // Invalid token
}

File: ios/SoulCalc/Domain/Parser/Token.swift
// SPDX-License-Identifier: MIT
// Copyright (c) 2026 Albert Zenkoff

import Foundation

/// Enumeration of all token types recognized by the lexer.
enum TokenType {
    // Literals
    case number         // 123, 45.67

    // Operators
    case plus           // +
    case minus          // - or −
    case multiply       // * or ×
    case divide         // / or ÷
    case percent        // %
    case power          // ^
    case equals         // =

    // Delimiters
    case lparen         // (
    case rparen         // )

    // Identifiers
    case variable       // $name (named variable)
    case lineRef        // $1, $2 (line reference)

    // Functions
    case sqrt           // sqrt

    // Special
    case eof            // End of input
    case error          // Invalid token
}

/// Represents a token produced by the lexer.
struct Token {
    let type: TokenType
    let value: String
    let position: Int
}
